### PR TITLE
Change PolicyManager interface

### LINKED ISSUE #700: Change PolicyManager interface
Change PolicyManager interface and the existing GrpcPolicyManager implementation, drop MultiPolicyManager. This deals with generalisation of Policy manager interface to adapt to opeanpis based format.

### LINKED ISSUE #701: Add OpenAPIPolicyManager implementation
This is a follow up to #700 
Add OpenAPIPolicyManager implementation based on the modified Policy Manager interface.

### PR DESCRIPTION
**Pull request by rohithdv**
This PR fixes #700 and #701 . 
This is a followup to #751 . This PR has the following changes:

1. Change to the PolicyManager interface to use openapi objects instead of grpc objects.
2. Changes to GRPC Policy manager implementation to account for the interface changes
3. Addition of conversion functions to/from GRPC objects from.to OpenApi objects. 
4. Changes in some places in manager test code to account for the interface change.
5. Addition of test code to `test/service/pilot `folder to test the above changes. 
6. Removal of `policymanager_multi` implementation which was accessing and combining results from multiple policy managers. Subsequent commenting of usage of `policymanager_multi` implementation in various places in the codebase.

As this PR leads to changes in the [mock policy manager](https://github.com/rohithdv/fybrik/blob/change_pm_interface/test/services/policymanager/serverpolicymanager.go), we have also provided a fix to implement opeapipolicy manager based on this new interface. Hence this PR also fixes #701 .

Signed-off-by: rohithdv <rovallam@in.ibm.com>

### CODE REVIEW

File: connectors/open_policy_agent/go/api_default.go [hunk 1]
```diff
@@ -13,12 +13,18 @@
...
+	"fmt"
+	"io/ioutil"
 	"net/http"
 
 	"github.com/gin-gonic/gin"
 )
 
 // GetPoliciesDecisions - getPoliciesDecisions
 func GetPoliciesDecisions(c *gin.Context) {
+	fmt.Println("in GetPoliciesDecisions of V2 OPA Connector!")
```
Comment (ronenkat) [inline connectors/open_policy_agent/go/api_default.go:25]: why are we using fmt.println/printf? where does it go? should this be log?

File: manager/controllers/mockup/policymanager.go [hunk 2]
```diff
@@ -19,7 +20,11 @@ type MockPolicyManager struct {
 }
 
 // GetPoliciesDecisions implements the PolicyCompiler interface
-func (s *MockPolicyManager) GetPoliciesDecisions(ctx context.Context, in *pb.ApplicationContext) (*pb.PoliciesDecisions, error) {
+// func (s *MockPolicyManager) GetPoliciesDecisions(ctx context.Context, in *pb.ApplicationContext) (*pb.PoliciesDecisions, error) {
```
Comment (ronenkat) [inline manager/controllers/mockup/policymanager.go:23]: why is this commented? remove or uncomment

File: manager/main.go [hunk 3]
```diff
@@ -245,23 +245,16 @@ func newPolicyManager() (connectors.PolicyManager, error) {
 	mainPolicyManagerName := os.Getenv("MAIN_POLICY_MANAGER_NAME")
 	mainPolicyManagerURL := os.Getenv("MAIN_POLICY_MANAGER_CONNECTOR_URL")
 	setupLog.Info("setting main policy manager client", "Name", mainPolicyManagerName, "URL", mainPolicyManagerURL, "Timeout", connectionTimeout)
-	policyManager, err := connectors.NewGrpcPolicyManager(mainPolicyManagerName, mainPolicyManagerURL, connectionTimeout)
+	// policyManager, err := connectors.NewGrpcPolicyManager(mainPolicyManagerName, mainPolicyManagerURL, connectionTimeout)
```
Comment (ronenkat) [inline manager/main.go:248]: pls remove comments

File: pkg/connectors/clients/policymanager_grpc.go [hunk 4]
```diff
@@ -39,11 +46,333 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
 	}, nil
 }
 
-func (m *grpcPolicyManager) GetPoliciesDecisions(ctx context.Context, in *pb.ApplicationContext) (*pb.PoliciesDecisions, error) {
-	result, err := m.client.GetPoliciesDecisions(ctx, in)
-	return result, errors.Wrap(err, fmt.Sprintf("get policies decisions from %s failed", m.name))
+func (m *grpcPolicyManager) GetPoliciesDecisions(
+	in *openapiclientmodels.PolicyManagerRequest, creds string) (*openapiclientmodels.PolicyManagerResponse, error) {
+	// log.Println("printing  stack trace in GetPoliciesDecisions")
```
Comment (ronenkat) [inline pkg/connectors/clients/policymanager_grpc.go:51]: remove comments

File: pkg/connectors/clients/policymanager_grpc.go [hunk 5]
```diff
@@ -39,11 +46,333 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
+	log.Println("GRPC result returned from GetPoliciesDecisions:", result)
+	policyManagerResp, _ := ConvertGrpcRespToOpenAPIResp(result)
+
+	res, err := json.MarshalIndent(policyManagerResp, "", "\t")
+	log.Println("err :", err)
+	log.Println("policyManagerResp: created from convGrpcRespToOpenApiResp")
+	log.Println("marshalled response:", string(res))
+	return policyManagerResp, nil
+
+	// result, err := m.client.GetPoliciesDecisions(ctx, in)
```
Comment (ronenkat) [inline pkg/connectors/clients/policymanager_grpc.go:69]: code comments....

File: pkg/connectors/clients/policymanager_grpc.go [hunk 6]
```diff
@@ -39,11 +46,333 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
+func ConvertOpenAPIRespToGrpcResp(
+	out *openapiclientmodels.PolicyManagerResponse,
+	datasetID string, op *pb.AccessOperation) (*pb.PoliciesDecisions, error) {
+	res, err := json.MarshalIndent(out, "", "\t")
+	log.Println("err :", err)
+	log.Println("Marshalled response in ConvertOpenAPIRespToGrpcResp:", string(res))
+
+	resultItems := out.GetResult()
+	enforcementActions := make([]*pb.EnforcementAction, 0)
+	usedPolicies := make([]*pb.Policy, 0)
```
Comment (ronenkat) [inline pkg/connectors/clients/policymanager_grpc.go:179]: is these policies or policy decisions? should we clarify if they are decisions?

File: test/services/pilot/mock-openapi-policymanager/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2020 IBM Corp.
+# SPDX-License-Identifier: Apache-2.0
+
+resources:
+- manager_client.yaml
+namespace: fybrik-system
+
+images:
+- name: manager-client
+  newName: uk.icr.io/datamesh/manager-client
```
Comment (ronenkat) [inline test/services/pilot/mock-openapi-policymanager/kustomization.yaml:10]: is this the right repo?

File: test/services/policymanager/serverpolicymanager.go [hunk 8]
```diff
@@ -4,33 +4,62 @@
...
 	PORT = 50082
 )
 
-func main() {
-	address := utils.ListeningAddress(PORT)
-	log.Printf("starting mock policy manager server on address %s", address)
+var router *gin.Engine
 
-	listener, err := net.Listen("tcp", address)
+func constructPolicyManagerRequest(inputString string) *openapiclientmodels.PolicyManagerRequest {
```
Comment (ronenkat) [inline test/services/policymanager/serverpolicymanager.go:25]: fmt? where does it go? is this the right output mechanism?

**Review by ronenkat — COMMENTED** [review]
please review my feedback as comments.

File: connectors/open_policy_agent/go/api_default.go (hunk 1 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go:25]: fixed to use log package. Please check. Thanks.

File: manager/controllers/mockup/policymanager.go (hunk 2 above)
Comment (rohithdv) [reply manager/controllers/mockup/policymanager.go:23]: Fixed. Thanks.

File: manager/main.go (hunk 3 above)
Comment (rohithdv) [reply manager/main.go:248]: fixed. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 4 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go:51]: fixed. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 5 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go:69]: fixed. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 6 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go:179]: usedPolicies are policies which when evaluated led to the corresponding enforcement actions.

File: test/services/pilot/mock-openapi-policymanager/kustomization.yaml (hunk 7 above)
Comment (rohithdv) [reply test/services/pilot/mock-openapi-policymanager/kustomization.yaml:10]: changed to ghcr.io. Thanks.

File: test/services/policymanager/serverpolicymanager.go (hunk 8 above)
Comment (rohithdv) [reply test/services/policymanager/serverpolicymanager.go:25]: changed to use log package. Thanks.

File: manager/controllers/app/policy_compiler_interface.go [hunk 9]
```diff
@@ -38,8 +38,16 @@ func ConstructApplicationContext(datasetID string, input *app.FybrikApplication,
...
+	log.Println("AccessOperation before calling ConstructApplicationContext in LookupPolicyDecisions", op)
 	appContext := ConstructApplicationContext(datasetID, input, op)
-	pcresponse, err := policyManager.GetPoliciesDecisions(context.Background(), appContext)
+	log.Println("appContext before calling GetPoliciesDecisions in LookupPolicyDecisions", appContext)
+	openapiReq, creds, _ := connectors.ConvertGrpcReqToOpenAPIReq(appContext)
+	log.Println("openapiReq after calling ConvertGrpcReqToOpenAPIReq in LookupPolicyDecisions", openapiReq)
+	openapiResp, err := policyManager.GetPoliciesDecisions(openapiReq, creds)
+	log.Println("openapiresp after GetPoliciesDecisions in LookupPolicyDecisions", openapiResp)
+	pcresponse, _ := connectors.ConvertOpenAPIRespToGrpcResp(openapiResp, datasetID, op)
+	log.Println("pcresponse after GetPoliciesDecisions in LookupPolicyDecisions", pcresponse)
```
Comment (roee88) [inline manager/controllers/app/policy_compiler_interface.go:49]: Too verbose, do we really need all of these?

File: manager/controllers/mockup/policymanager.go [hunk 10]
```diff
@@ -94,6 +98,15 @@ func (s *MockPolicyManager) GetPoliciesDecisions(ctx context.Context, in *pb.App
...
-		DatasetDecisions: dataSetWithActions}, nil
+	result := &pb.PoliciesDecisions{ComponentVersions: externalComponents,
+		DatasetDecisions: dataSetWithActions}
+
+	policyManagerResp, _ := connectors.ConvertGrpcRespToOpenAPIResp(result)
+
+	res, err := json.MarshalIndent(policyManagerResp, "", "\t")
+	log.Println("err :", err)
+	log.Println("policyManagerResp: created from convGrpcRespToOpenApiResp")
+	log.Println("marshalled response:", string(res))
```
Comment (roee88) [inline manager/controllers/mockup/policymanager.go:109]: Why are you printing the error and then ignoring it? shouldn't it just be returned?

File: pkg/connectors/clients/policymanager_grpc.go [hunk 11]
```diff
@@ -39,11 +46,327 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
-func (m *grpcPolicyManager) GetPoliciesDecisions(ctx context.Context, in *pb.ApplicationContext) (*pb.PoliciesDecisions, error) {
-	result, err := m.client.GetPoliciesDecisions(ctx, in)
-	return result, errors.Wrap(err, fmt.Sprintf("get policies decisions from %s failed", m.name))
+func (m *grpcPolicyManager) GetPoliciesDecisions(
+	in *openapiclientmodels.PolicyManagerRequest, creds string) (*openapiclientmodels.PolicyManagerResponse, error) {
+	log.Println("openapiclientmodels.PolicyManagerRequest: received in GetPoliciesDecisions: ", *in)
+	appContext, _ := ConvertOpenAPIReqToGrpcReq(in, creds)
+	log.Println("appContext: created from convertOpenApiReqToGrpcReq: ", appContext)
+
+	result, _ := m.client.GetPoliciesDecisions(context.Background(), appContext)
```
Comment (roee88) [inline pkg/connectors/clients/policymanager_grpc.go:55]: Ignoring error?

File: pkg/connectors/clients/policymanager_grpc.go [hunk 12]
```diff
@@ -39,11 +46,327 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
+
+	log.Println("GRPC result returned from GetPoliciesDecisions:", result)
+	policyManagerResp, _ := ConvertGrpcRespToOpenAPIResp(result)
```
Comment (roee88) [inline pkg/connectors/clients/policymanager_grpc.go:58]: Ignoring error?

File: pkg/connectors/clients/policymanager_grpc.go [hunk 13]
```diff
@@ -39,11 +46,327 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
+
+	res, err := json.MarshalIndent(policyManagerResp, "", "\t")
+	log.Println("err :", err)
+	log.Println("policyManagerResp: created from convGrpcRespToOpenApiResp")
+	log.Println("marshalled response:", string(res))
+	return policyManagerResp, nil
```
Comment (roee88) [inline pkg/connectors/clients/policymanager_grpc.go:64]: same comment on ignoring error

File: pkg/connectors/clients/policymanager_grpc.go [hunk 14]
```diff
@@ -39,11 +46,327 @@ func NewGrpcPolicyManager(name string, connectionURL string, connectionTimeout t
...
 }
+
+// ref: https://sosedoff.com/2014/12/15/generate-random-hex-string-in-go.html
+func randomHex(n int) (string, error) {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
```
Comment (roee88) [inline pkg/connectors/clients/policymanager_grpc.go:78]: Functions that might be generally useful should go to their own package in `pkg/`

File: manager/controllers/app/policy_compiler_interface.go (hunk 9 above)
Comment (rohithdv) [reply manager/controllers/app/policy_compiler_interface.go:49]: Modified the logs to remove unnecessary logs and modified log messages also. Please review. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 13 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go]: Fixed. Please review. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 12 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go]: Fixed. Please review. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 11 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go]: Fixed. Please review. Thanks.

File: manager/controllers/mockup/policymanager.go (hunk 10 above)
Comment (rohithdv) [reply manager/controllers/mockup/policymanager.go]: Fixed. Please review. Thanks.

File: pkg/connectors/clients/policymanager_grpc.go (hunk 14 above)
Comment (rohithdv) [reply pkg/connectors/clients/policymanager_grpc.go]: Added a `utils` folder in `pkg/`. Please review. Thanks.

**Review by roee88 — APPROVED** [review]

**Comment by roee88** [comment]
We will do a round on improved logging after we finish integration

**Pull request merged**